### ISSUE TITLE
Webpage capture boefje is broken

### ISSUE THREAD

**Issue by ehasekamp** [SSC-ICT-Innovatie/nl-kat-coordination#3916]
`bug`, `boefjes`,`webpage capture boef`

**Describe the bug**
The webpage capture boefje is broken. Might be related to #1190

**To Reproduce**

1. Go to KAT-alogus --> Webpage Capture
2. Click on Start Scan

**Expected behavior**
No Failed scan

**Screenshots**

024-11-28T19:50:08.861231058Z npm warn exec The following package was not found and will be installed: playwright@1.49.0
2024-11-28T19:50:10.770969195Z Error: t.parse: Executable doesn't exist at /ms-playwright/chromium_headless_shell-1148/chrome-linux/headless_shell
2024-11-28T19:50:10.770986205Z ╔══════════════════════════════════════════════════════════════════════╗
2024-11-28T19:50:10.770990795Z ║ Looks like Playwright Test or Playwright was just updated to 1.49.0. ║
2024-11-28T19:50:10.770993925Z ║ Please update docker image as well.                                  ║
2024-11-28T19:50:10.770996835Z ║ -  current: mcr.microsoft.com/playwright:v1.46.1-jammy               ║
2024-11-28T19:50:10.770999655Z ║ - required: mcr.microsoft.com/playwright:v1.49.0-jammy               ║
2024-11-28T19:50:10.771002546Z ║                                                                      ║
2024-11-28T19:50:10.771005696Z ║ <3 Playwright Team                                                   ║
2024-11-28T19:50:10.771009216Z ╚══════════════════════════════════════════════════════════════════════╝
2024-11-28T19:50:10.779776181Z npm notice
2024-11-28T19:50:10.779796441Z npm notice New minor version of npm available! 10.8.1 -> 10.9.1
2024-11-28T19:50:10.779801902Z npm notice Changelog: https://github.com/npm/cli/releases/tag/v10.9.1
2024-11-28T19:50:10.779806052Z npm notice To update run: npm install -g npm@10.9.1
2024-11-28T19:50:10.779809872Z npm notice

**OpenKAT version**
`v1.17.0` and `0.0.1.dev1` main branch

**Desktop (please complete the following information if relevant):**

- OS: macOS
- Browser: Chrome
- Version: 

**Additional context**

**Comment by zcrt** [comment]
Looks like the `latest` tag is no longer used by Playwright: https://github.com/microsoft/playwright/releases/tag/v1.47.0, https://github.com/microsoft/playwright/issues/32483. If I remember correctly, version pinning gave problems in the past as playwright tries to auto-update.

I would suggest making the docker tag a boefje variable, so that the end-user may be able to fix some stuff whenever Microsoft decides that something unexpected needs to be done. Not sure how to handle this case by default though.

**Comment by hasecon** [comment]
Picking this back up — still broken, and it's the same root cause as the 2024 report, just with newer version numbers. Re-confirmed live on 2026-08-18 (every `webpage-capture` task's container exits 1):

```
npm warn exec The following package was not found and will be installed: playwright@1.62.1
Error: Executable doesn't exist at /ms-playwright/chromium_headless_shell-1234/chrome-linux/headless_shell
current:  mcr.microsoft.com/playwright:v1.53.0-noble
required: mcr.microsoft.com/playwright:v1.62.1-noble
```

### Root cause

There are two unpinned points, and the runtime one is the time-bomb:

1. **Runtime:** `main.py` runs `["/usr/bin/npx", "playwright", "screenshot", …]`. No playwright package is installed in the image, so `npx` fetches **the latest** playwright from npm at run time. The base image's baked browsers are for the pinned base version (1.53), but latest-playwright wants a newer browser revision → crash. This re-breaks on **every** Playwright release, regardless of when the image was built — which is exactly why it broke in 2024 (v1.46→1.49) and again now (v1.53→1.62).
2. **Build:** `RUN npx playwright install --with-deps chromium` is also unpinned.

@zcrt's earlier point stands — Playwright dropped the `latest` docker tag and it auto-updates — but the fix is precisely to stop letting it auto-resolve: pin the package to the base image and never `npx`-fetch at runtime.

### Proposed fix (durable)

**1. Pin the playwright package to the base image, and don't use `npx` at runtime.** One version drives both the base tag and the npm package, so browsers and package can never drift. Same shape as the `NUCLEI_VERSION` ARG in `kat_nuclei_cve`:

```dockerfile
ARG PLAYWRIGHT_VERSION=1.53.0
# tag for readability + digest for immutability (also addresses #3261)
FROM mcr.microsoft.com/playwright:v1.53.0-noble@sha256:<resolved-digest>
...
RUN npm install -g playwright@${PLAYWRIGHT_VERSION} \
 && playwright install --with-deps chromium   # browsers already baked in the matching base
```

and in `main.py`, invoke the pinned binary instead of `npx`:

```python
command = ["playwright", "screenshot", "-b", browser, "--full-page", "--ignore-https-errors", ...]
```

**2. Pin the base by digest (A2 / #3261)** as shown above — `:v1.53.0-noble@sha256:…`. Keeps the human-readable tag *and* an immutable digest; Dependabot updates the digest while keeping the tag.

### Keeping it from silently rotting again

The reason this sat broken for ~21 months is that **nothing exercises webpage-capture in CI**, so a bad version never fails a build. Two guardrails:

- **Dependabot on this Dockerfile** (docker ecosystem) so the pin/digest are bumped deliberately, with CI, instead of rotting. Dependabot already runs in this repo.
- **A CI smoke test for capture** — minimally assert `playwright --version` matches the baked browser (`/ms-playwright/chromium-*`), ideally a tiny real capture. This catches a mismatched bump before merge and would have caught both the 2024 and 2026 breakage.

### Scope note

This fix is just "make capture work again + keep it working". The related **#1559** (feed Playwright an `HTTPResource` + specific IP instead of a bare `HostnameHTTPURL`) is a separate improvement I'd fold into follow-up work on the capture flow, not this fix.

@underdarknl — if the A1+A2 shape above (pinned package via ARG, `FROM tag@digest`, call the binary not `npx`, plus the CI smoke test) looks right, I'll open the PR that closes this. It also unblocks running Wappalyzer over the browser HAR (the tech/version-detection gap in #4447), which needs capture producing a HAR first.

**Comment by underdarknl** [comment]
Ack, pinning the version is preferred, directly calling the playwright executable seems like a valid route.

With regards to #4447 fixing the har-file production using playwright works for wappalyzer specifically because playwright does not adhere to our scan levels with regards to referenced other hosts and urls (possibly behind redirects). We should probably make sure we communicatie this breach of our own rules.

**Comment by hasecon** [comment]
Thanks — PR incoming with exactly that: `PLAYWRIGHT_VERSION` arg driving the pinned base + `npm install -g playwright@<ver>`, `main.py` calling the `playwright` binary directly, base pinned by digest, a Dependabot docker entry, and a build-time `playwright --version` guard (the capture image is built in `containerized_boefjes`, so the guard runs in CI).

Good call on the scan-level point, and agreed it's important. To be clear it's **not introduced by this fix** — Playwright has always rendered the page and pulled in whatever it references (third-party scripts, CDNs, redirect targets), none of which pass through our per-OOI scan-level/clearance checks. The fix just makes capture run again, so the behaviour becomes visible again.

Where it actually bites is the **#4447 follow-up** (turning the browser HAR into `Software`/OOIs): that would mint OOIs for hosts/URLs the user never cleared — a real breach of our own rules, not just extra network traffic. So I'll:
- document the caveat on the capture boefje itself (it fetches referenced third-party resources outside OpenKAT's scan-level model), and
- carry it explicitly into the Niveau-1 design — e.g. only derive OOIs for resources on already-known/cleared hosts, or clearly mark browser-sourced detections as lower-confidence/out-of-scan-scope.

I'll keep this fix PR to just "capture works + stays working" and raise the scan-level handling as its own design point on the #4447 line so it gets the attention it deserves.
